Initial implementation of the RelationshipManager system.

permcody opened the pull request:
More work needs to be done to remove the existing ghosting methods
out of MooseMesh.

refs #10455

RelationshipManagers are wrappers for libMesh ghosting_functors. They are made to be created AND executed during the ActionWarehouse traversal (i.e. way early in MOOSE setup) so that they can be attached to the corresponding libMesh objects at the right times.

Along with this PR, I have implemented the first two RelationshipManagers (ElementSideNeighborLayers and ElementPointNeighbors) which make both of those "entities" available when using distributed mesh.

permcody committed:
Initial implementation of the RelationshipManager system.
More work needs to be done to remove the existing ghosting methods
out of MooseMesh.

refs #10455

mangerij commented:
Will this be useful in implementing gradient periodicity?

permcody commented:
Probably. If we do this correctly, any geometric, or solution resource that's non local can be made available to you. I haven't looked at, or thought about this problem long enough to know what information we'll need beyond what we have in value periodic boundary conditions. Interestingly, we could redo how PBCs work in general to use this more general system (among other things).

moosebuild commented:
Job [Documentation](https://www.moosebuild.org/job/145041/) on dec1470 wanted to post the following:

View the site [here](http://mooseframework.org/docs/PRs/10506/site)

This comment will be updated on new commits.

friedmud commented:
Great work Cody!

I’ll try to look at the code tomorrow.
On Thu, Jan 11, 2018 at 4:41 PM MOOSE Build <notifications@github.com>
wrote:

> Job Documentation <https://www.moosebuild.org/job/145041/> on dec1470
> <https://github.com/idaholab/moose/commit/dec14707c997a659505ff0237cbbb70565d6c976>
> wanted to post the following:
>
> View the site here <http://mooseframework.org/docs/PRs/10506/site>
>
> This comment will be updated on new commits.
>
> —
> You are receiving this because your review was requested.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/10506#issuecomment-357097881>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AA1JMdJAX5ow5y-Q9QyN_VBS8lVNHpJdks5tJpw9gaJpZM4Rbj5k>
> .
>

permcody commented:
Still waiting for that review :)

permcody committed:
Reworking the initial design

permcody commented:
@friedmud - This is getting pretty close as far as the design goes. I have several more loose ends to cleanup:

- Documentation (both on new methods and adding pages to MooseDocs)
- Algebraic Test cases
- Calling reinit on the mesh one more time for late deletes with test

None of that should impact the design here.

jwpeterson commented on framework/include/utils/MooseTypes.h:
```diff
@@ -23,6 +23,10 @@
+#include "libmesh/parameters.h"
+
+// BOOST include
+#include "bitmask_operators.h"
```
Do you remember what version of Boost this header originally came from? It seems very useful but I don't see it in Boost versions going back to 1.61.

permcody replied:
Don't know, but it's been in libMesh for awhile. I've been using it in the GrainTracker since early last year before we added BOOST to the package.

permcody replied:
OK - I'm senile. I added this to MOOSE

jwpeterson commented on framework/src/actions/ActionWarehouse.C:
```diff
@@ -200,21 +201,16 @@ ActionWarehouse::buildBuildableActions(const std::string & task)
-    for (std::multimap<std::string, std::string>::const_iterator it = range.first;
-         it != range.second;
-         ++it)
+    auto it_pair = _action_factory.getActionsByTask(task);
```
:+1: C++11 version is a major improvement!

jwpeterson commented on framework/src/actions/AddRelationshipManager.C:
```diff
@@ -0,0 +1,39 @@
+void
+AddRelationshipManager::act()
+{
+  Moose::RelationshipManagerType rm_type = _current_task == "add_geometric_rm"
```
I know the order of operations is fine here, but you might consider adding parentheses around the ternary statement for readability.

jwpeterson commented on framework/src/base/Coupleable.C:
```diff
@@ -29,7 +29,6 @@ Coupleable::Coupleable(const MooseObject * moose_object, bool nodal)
     _c_is_implicit(_c_parameters.have_parameter<bool>("implicit")
                        ? _c_parameters.get<bool>("implicit")
                        : true),
-    _coupleable_params(_c_parameters),
```
That little guy? Don't worry about him.

jwpeterson commented on framework/src/base/LazyCoupleable.C:
```diff
@@ -0,0 +1,64 @@
+    _coupled_var_numbers[var_name] = libmesh_make_unique<unsigned int>(0);
+}
+
+LazyCoupleable::~LazyCoupleable() {}
```
We omit empty destructors now, right?

permcody replied:
Yes and No, you need them for the BASE CLASS, but not any derived classes. However, looking at this again. I don't intend for this class to be derived from so I'm going to remove it.

jwpeterson commented on framework/src/relationshipmanagers/RelationshipManager.C:
```diff
@@ -0,0 +1,127 @@
+
+  /**
+   * This parameter indicates which type of RelationshipManager this class represents prior to
+   * its
```
This comment got reflowed strangely by clang-format.

jwpeterson commented on test/tests/mesh/ghost_functors/tests:
```diff
@@ -12,7 +12,7 @@
     input = 'geometric_neighbors.i'
-    cli_args = 'Mesh/num_ghosted_layers=2 Outputs/file_base=geometric_edge_two_2D_out'
+    cli_args = 'UserObjects/ghost_uo/element_side_neighbor_layers=2 Outputs/file_base=geometric_edge_two_2D_out'
     exodiff = 'geometric_edge_two_2D_out.e'
```
What is the exodiff actually testing for with regard to the number of ghosted layers in these tests?

permcody replied:
AuxVariable that contains the union of all processors ghosted elements.

permcody commented on framework/include/base/LazyCoupleable.h:
```diff
@@ -0,0 +1,68 @@
+   */
+  LazyCoupleable(const MooseObject * moose_object);
+
+  ~LazyCoupleable();
```
This is an Interface class. I do not intend for it to be derived from. I'll add `final` to the class declaration.

permcody commented on framework/include/mesh/MooseMesh.h:
```diff
@@ -449,6 +450,14 @@ class MooseMesh : public MooseObject, public Restartable
 
+  /**
+   * Method for adding Relationship managers (libMesh ghosting functors) to the Mesh object. The
+   * parameter is passed in via shared_ptr. It's expected that the Mesh object will assume ownership
```
Stale method (the body even got removed). The App will hold onto all GhostingFunctors since  libMesh doesn't currently have a mechanism to hold unto them directly. libMesh expects a reference and the lifetime needs to last for the lifetime of the libMesh objects to which they are attached.

permcody commented on framework/include/relationshipmanagers/RelationshipManager.h:
```diff
@@ -0,0 +1,93 @@
+protected:
+  /**
+   * Method to setup the RelationshipManager for use in the simulation. It is called exactly
+   * once for each type of functor object that this RelationshipManager is capable of creating.
```
No, "at most once" would be more correct. I'll see if I can clean this up a bit more.

permcody committed:
Address comments

Also, I fixed several const correctness issues I ran into when getting
LazyCoupleable to compile
